File: java/src/singleton/Singleton.java
package singleton;

public class Singleton {
    private volatile static Singleton singleton;

    private Singleton() {
    }

    //采用双锁机制，安全且在多线程情况下能保持高性能
    public static Singleton getInstance() {
        if (singleton == null) {
            synchronized (Singleton.class) {
                if (singleton == null) {
                    singleton = new Singleton();
                }
            }
        }

        return singleton;
    }


}

File: golang/singleton/singleton.go
package singleton

import "sync"

type Singleton struct{}

var singleton *Singleton
var once sync.Once

//使用once.Do保证同一时间只能有一个线程操作变量
//此处也可以使用sync.Mutex来加锁，但是没有once.Do优雅
func GetInstance() *Singleton {
	once.Do(func() {
		singleton = &Singleton{}
	})
	return singleton
}
